### PR TITLE
[Issue-1260] - Github action to generate pydantic models

### PR DESCRIPTION
**Pull request by pmbrull**
### Describe your changes :

This PR fixes https://github.com/open-metadata/OpenMetadata/issues/1260.

We currently have the make generate recipe taking care of generating the models. However, it is easy to miss updates and get code not properly aligned.

The prepared action checks all pushes against `main` that update code inside `catalog-rest-service/src/main/resources/json/**`, so it will only run if we are indeed changing JSON schemas.

Then, it will run the `make generate` recipe and push the code to the repo with a fake user `open-metadata`.

Here we might need to check if this is the right approach or if we should have a bot account with proper write permissions.

> OBS: It also fixes a random missed import that breaks a test

Thanks!

#
### Type of change :

- [ ] Bug fix
- [ ] Improvement
- [x] New feature
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/open-source-community/developer) document.
- [ ] I have performed a self-review of my own. 
- [ ] I have tagged my reviewers below.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] All new and existing tests passed.

#
### Reviewers
@sureshms @harshach, if you could take a look, please! 🙏  Any thoughts on the permissions?

Thanks!

### CODE REVIEW

File: .github/workflows/py-generate.yml
Comment (mithmatt) [inline .github/workflows/py-generate.yml:14]: When does this action run ?

While CI checks out main branch?

Comment (pmbrull) [reply .github/workflows/py-generate.yml:14]: Hi, @mithmatt,

At the top of the action we have defined
```
on:
  push:
    branches:
      - main
    paths:
      - 'catalog-rest-service/src/main/resources/json/**'
```

So it will run after a push to `main` if there are updated files under the JSON schemas dir. This way we can be sure that the origin's pydantic models are up to date with any update on the JSON schemas.

**Comment by pmbrull** [comment]
I'll merge this after checking https://github.com/open-metadata/OpenMetadata/issues/1268 to be sure the generated files are consistent. Thanks!

File: .github/workflows/py-generate.yml
Comment (mithmatt) [reply .github/workflows/py-generate.yml:14]: Thanks for clarifying that. 

If I understand correctly, following are the sequence of events:

1. Developer changes json schema, writes and commits code (commit A)
2. Committer merges code into main branch
3. Githhub workflow kicks in
4. Github creates another commit (commit B) with pydantic models


Should pydantic models be generated as part of commit A itself, so that they go in as one single commit?
I think a client side git [pre-commit](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks) hook is more appropriate for this use case.

Comment (pmbrull) [reply .github/workflows/py-generate.yml:14]: Yep exactly, that is the workflow.

I also thought about coupling/decoupling the commits and while it might feel cleaner to have it in a single shot, I also believe it is important that the process acts in an independent and reproducible way. As we want to encourage different collaborators to play around with the schemas, adding an extra layer of complexity where they need to have the right environment and packages ready might be a bit too much.

I feel it might be more valuable to have a simple process that we can easily track even if that means having a specific commit where the codebase won't be aligned.

WDYT?

> Also that might me remember that I need to install the python dependencies in the action!

Comment (mithmatt) [reply .github/workflows/py-generate.yml:14]: I see we are trading off correctness for simplicity + developer velocity with the github workflow. 
I agree it is better to have a simple process.

For jsonschema to java pojos - mvn generate-sources (runs as part of mvn test) ensures that the classes can be generated successfully when a pull request is raised.

Should we have a similar one for pydantic models (to prevent us from checking in code where this generation could potentially fail)?

PS: I'm probably extra cautious about breaking something.

Comment (pmbrull) [reply .github/workflows/py-generate.yml:14]: actually, you have a good point, as not that long ago we needed to slightly update a JSON schema change as the pydantic generation failed.

As part of the CI process, we could have a separate job that validates this conversion, even if it does not update the current PR. I also would advise against validating and generating the sources during the PR as the review becomes quite cumbersome, with a lot of changes we do not even want to see.

But yep, +1 on being extra cautious here! I can prepare a similar workflow to run with each PR. Let's try to keep the automation as safe as possible.

**Review by mithmatt — APPROVED** [review]

**Comment by pmbrull** [comment]
Hi, @mithmatt,

Thanks for your input on this, really appreciate the chance to discuss it out loud!

I have made the last change to include what we were discussing. The action will now run for both PRs and Push events (considering the JSON schemas path was updated).

However, the models will be pushed only in the case of a `push` event:

```
- name: Push models
        # Push the generated models if we are running in a push to main
        if: ${{ github.event_name == 'push' }}
        run: |
          git config --global user.name 'open-metadata'
         [...]
```

Otherwise, the process will stop at `make generate`, which will help us identify if the schema changes are able to run the job properly.

Thanks!

**Comment by harshach** [comment]
Thanks @pmbrull. One of the options we are thinking is removing generated code from the repo and build it when we do pip package and for any dev purpose. We already have the make recipe thanks to you, we can use that and document that as prerequisite for development purpose.
We can also make recipe for publishing the package which will generate the code and publish the package cc @ayush-shah .
What do you think?

**Comment by ayush-shah** [comment]
> Thanks @pmbrull. One of the options we are thinking is removing generated code from the repo and build it when we do pip package and for any dev purpose. We already have the make recipe thanks to you, we can use that and document that as prerequisite for development purpose. We can also make recipe for publishing the package which will generate the code and publish the package cc @ayush-shah . What do you think?

It will be more clean and efficient that way and no need to worry about generating pydantic models everytime we publish a code. Sounds good to me @harshach.

**Comment by pmbrull** [comment]
Makes sense @harshach, will be more consistent with the backend approach as well. Let's review the package generation to add this.

I'll simplify the action to not push the models but to validate that changes in the JSON schema will not break the pydantic generation

**Comment by pmbrull** [comment]
Merging now, I'll open a follow-up PR for https://github.com/open-metadata/OpenMetadata/issues/1268 which will help in validating this new action. Thanks!

**Pull request merged**
